bluemira/equilibria/coils.py

```diff
@@ -22,42 +22,74 @@
 """
 Coil and coil grouping objects
 """
+from __future__ import annotations
 
-from copy import deepcopy
-from re import split
-from typing import Any, Optional
+import abc
 
-import matplotlib.pyplot as plt
+# from copy import deepcopy
+from enum import Enum, EnumMeta, auto
+from functools import update_wrapper
+
+# from re import split
+from typing import Callable, Dict, Iterable, List, Optional, Tuple, Union
+
+# import matplotlib.pyplot as plt
 import numpy as np
-from scipy.interpolate import RectBivariateSpline
 
 from bluemira.base.constants import MU_0
-from bluemira.base.look_and_feel import bluemira_warn
-from bluemira.equilibria.constants import (
-    I_MIN,
-    J_TOR_MIN,
+
+# from bluemira.base.look_and_feel import bluemira_warn
+from bluemira.equilibria.constants import (  # I_MIN,; J_TOR_MIN,
     NBTI_B_MAX,
     NBTI_J_MAX,
     X_TOLERANCE,
 )
 from bluemira.equilibria.error import EquilibriaError
-from bluemira.equilibria.file import EQDSKInterface
-from bluemira.equilibria.plotting import CoilPlotter, CoilSetPlotter, PlasmaCoilPlotter
-from bluemira.magnetostatics.greens import (
-    circular_coil_inductance_elliptic,
+
+# from bluemira.equilibria.file import EQDSKInterface
+# from bluemira.equilibria.plotting import CoilPlotter, CoilSetPlotter, PlasmaCoilPlotter
+from bluemira.magnetostatics.greens import (  # circular_coil_inductance_elliptic,
     greens_Bx,
     greens_Bz,
     greens_psi,
 )
 from bluemira.magnetostatics.semianalytic_2d import semianalytic_Bx, semianalytic_Bz
-from bluemira.utilities.tools import is_num
+from bluemira.utilities.tools import is_num_array
+
+# from scipy.interpolate import RectBivariateSpline
+
+
+__all__ = ["CoilType", "Coil", "CoilSet", "Circuit", "SymmetricCircuit"]
+
+
+class CoilTypeEnumMeta(EnumMeta):
+    """
+    Allow override of KeyError error string
+    """
+
+    def __getitem__(self, name):
+        try:
+            return super().__getitem__(name)
+        except KeyError:
+            raise KeyError(f"Unknown CoilType {name}") from None
+
+
+class CoilType(Enum, metaclass=CoilTypeEnumMeta):
+    """
+    CoilType Enum
+    """
+
+    PF = auto()
+    CS = auto()
+    NONE = auto()
 
-PF_COIL_NAME = "PF_{}"
-CS_COIL_NAME = "CS_{}"
-NO_COIL_NAME = "Unclassified_{}"
 
+__ITERABLE_FLOAT = Union[float, Iterable[float]]
+__ITERABLE_COILTYPE = Union[str, CoilType, Iterable[Union[str, CoilType]]]
+__ANY_ITERABLE = Union[__ITERABLE_COILTYPE, __ITERABLE_FLOAT]
 
-class CoilNamer:
+
+class CoilNumber:
```

Comment:
This is quite minor, but I've found that if you initialise a Coil incorrectly, it still increments the counters. This was probably the case before too, so we can forget it for now.

Reply:
I think its important to keep this. That way dont have to do a lot of catching in order to have unique coil numbers even if we don't necessarily have unique coil names.

I havent yet done it but maybe in the next PR the region and path links should be done with the coil numbers (also anything else that internally uses the names for access)